src/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs

```diff
@@ -298,9 +298,13 @@ private bool StartCore(ProcessStartInfo startInfo)
             bool setCredentials = !string.IsNullOrEmpty(startInfo.UserName);
             uint userId = 0;
             uint groupId = 0;
+            uint[] groups = null;
             if (setCredentials)
             {
-                (userId, groupId) = GetUserAndGroupIds(startInfo);
+                (userId, groupId, groups) = GetUserAndGroupIds(startInfo);
+                // We can get permission errors when setting credentials to ourself,
+                // so we only set credentials when we need to run as a different user.
+                setCredentials = userId != Interop.Sys.GetUid();
```

Comment:
Could we also avoid paying to get the groups information if !setCredentials, or do we still need it in that case?

Reply:
What happens in the unlikely event when a user changes their primary group or is added to a new group? E.g. a user starts their long running program which checks every 5 minutes if the user has access to a file using a command line script.

In this case the new created process will still have the old groups and gids because of this check.

As far as I remember when changing your primary group or group membership a newly created process in Linux will run with the new assigned groups.

I don't know if this is might be an unexpected behavior?

Reply:
What happens in the unlikely event when a user changes their primary group or is added to a new group? E.g. a user starts their long running program which checks every 5 minutes if the user has access to a file using a command line script.

In this case the new created process will still have the old groups and gids because of this check.

As far as I remember when changing your primary group or group membership a newly created process in Linux will run with the new assigned groups.

I don't know if this is might be an unexpected behavior?

Reply:
> Could we also avoid paying to get the groups information if !setCredentials, or do we still need it in that case?

We can avoid it.

> What happens in the unlikely event when a user changes their primary group or is added to a new group?

> As far as I remember when changing your primary group or group membership a newly created process in Linux will run with the new assigned groups.

> I don't know if this is might be an unexpected behavior?

Newly created processes inherit this from their parent, they are not aware group changes in the user database. So this is 'default' behaviour.
This check is here because without it, setting username to the current user fails on setgroups with EPERM.
We can move the check down to pal_process and ignore that EPERM.

Reply:
> We can move the check down to pal_process and ignore that EPERM.

I'm actually not in favour of this, because it makes the behaviour dependent whether the user has this capability or not. So, I'll stick to the check as is.